src/views/tutorial-view/utils/use-progress-toast.ts

```diff
@@ -0,0 +1,107 @@
+import { useEffect } from 'react'
+import { useCollectionProgress, useTutorialProgress } from 'hooks/progress'
+import usePrevious from 'hooks/use-previous'
+import { Collection, Tutorial } from 'lib/learn-client/types'
+import makeProgressToast from './make-progress-toast'
+
+/**
+ * Given a tutorialId, collectionId, and count of tutorials in the collection,
+ *
+ * Show progress toast when the specified tutorial changes progress,
+ * from some non-"complete" status to "complete".
+ *
+ * This toast will give information related to collection progress,
+ * which is why we need the array of tutorial IDs in the collection.
+ */
+export function useProgressToast({
```

Comment:
No action needed immediately here, just something to noodle on, but I would expect that the implementation of the toasts _might_ be simpler if we triggered them in our mutation `onSuccess` callbacks. 💭 In the effect here we're tracking the transition of the status value to determine when progress was updated, where the `onSuccess` callback would be the "definitive" point at which we know the state changes.

Reply:
Sweet, I dig this 👍 I guess we'd need to access collection progress in the mutation `onSuccess` as well though, right? 🤔 Will put my noodle cap on during optimization mode... made a ticket for later: 🎟️ [[progress toast] maybe simplify implementation by triggering toasts from mutation onSuccess callback](https://app.asana.com/0/1202097197789424/1202938785223946/f)

Reply:
Ah that is true 🤔 Now I'm wondering if the API could just return that information for us as part of the `POST` response 💭

Anyways, totally agree we can revisit at a later date!

Reply:
I think we can and should make the tutorial complete toasts render `onSuccess` of mutations. We show some toasts for bookmark interactions in this way.

Collection complete will be a little more challenging since we're not mutating that data and only fetching it. For that, I'd suggest updating `useCollectionProgress` to accept an `onSuccess` option that gets passed as an option to its `useQuery` and rendering the collection complete toast from there based on the query's data. This approach _should_ remove the need for checking several different pieces of state.

Nothing to do here though! This works and we can come back when there's time. ✨

Reply:
@ashleemboyer Sounds good!

I think there are definitely some challenges here... I think even if we used `useCollectionProgress` `onSuccess` to trigger the `"Collection complete"` toast, we'd still need to know collection progress wherever we trigger the other `"Tutorial complete, ? tutorials left"` (as we need the count, and we don't want to render a `"0 tutorials left"` toast).

Very on board with investigating further later, in 🎟️ [[progress toast] maybe simplify implementation by triggering toasts from mutation onSuccess callback](https://app.asana.com/0/1202097197789424/1202938785223946/f) 👍